### PR TITLE
Feat(plugins): Automatic requirements check

### PR DESCRIPTION
**Pull request by gmuloc**
## Change Summary

This PR intends to implement a plugin to verify that the minimum requirements for AVD are met

## Component(s) name

`arista.avd.plugins`

## Proposed changes

* Creation of a plugin
* Addition of task in roles to verify the environment
 
## How to test

* The initial commit for this PR can be checked with any playbook importing the `eos_designs` role by running with the `debug` tags.

* some tests will need to be written for the python module.

### User Checklist

[ ] decide when / where to run the task

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/plugins/action/verify_requirements.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import re
+import sys
+
+from ansible.plugins.action import ActionBase
+from ansible.errors import AnsibleActionFail
+from ansible.utils.display import Display
+
+HAS_PIP = True
+try:
+    from pip._vendor.packaging.specifiers import SpecifierSet
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/plugins/action/verify_requirements.py]: Please update requirement.txt to get pip in a min version that has this class

File: ansible_collections/arista/avd/plugins/modules/verify_requirements.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    required: True
+    type: list
+"""
+
+EXAMPLES = r"""
+- name: Verify python requirements
+  arista.avd.verify_requirements:
+    dependencies:
+      - jinja2 >= 2.9
+      - ansible == 2.13
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/plugins/modules/verify_requirements.py]: let's not use ansible == 2.13 as an example. Either `ansible-core` or something else than Ansible.

File: ansible_collections/arista/avd/roles/eos_designs/tasks/main.yml [hunk 3]
```diff
@@ -11,6 +11,20 @@
...
+- name: "Verify Requirements"
+  tags: [debug, avd_req]
+  arista.avd.verify_requirements:
+    dependencies: "{{ lookup('file', requirements_path ~ '/requirements.txt').splitlines() }}"
+  vars:
+    requirements_path: "{{ (role_path | split('/'))[0:-2] | join('/') }}"
+  run_once: true
+  register: requirements
+
+- name: "Print requirements"
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/tasks/main.yml]: please make the tags for this task `debug, avd_req, never`

**Review by ClausHolbechArista — COMMENTED** [review]
Looking good. We could consider if we could add the collection version stuff into this task(plugin, to avoid more housekeeping tasks.

File: ansible_collections/arista/avd/roles/eos_designs/tasks/main.yml (hunk 3 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/roles/eos_designs/tasks/main.yml]: MAde it for both tasks

File: ansible_collections/arista/avd/plugins/modules/verify_requirements.py (hunk 2 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/plugins/modules/verify_requirements.py]: using paramikoin next commit

File: ansible_collections/arista/avd/plugins/action/verify_requirements.py (hunk 1 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/plugins/action/verify_requirements.py]: Took the latest change to the class and used this as minimum version (21.3): https://github.com/pypa/pip/blame/main/src/pip/_vendor/packaging/specifiers.py

**Comment by gmuloc** [comment]
For the failing CI, it looks like we are hitting this: https://github.com/ansible/ansible/issues/71602 - the symptom is the same. The issue is that pylint itself is getting the warnings so I cannot do anything inside our own code. I tested locally with:

* Python 3.10.3
* virtualenv 20.13.4
* setuptools 65.3.0

```python
"""
test.py
"""
try:
    from pip._vendor.packaging.specifiers import SpecifierSet
except ImportError:
    print("blah")

s = SpecifierSet()
print(f"hello world {s}")

```

```
blah% python  -m pylint test.py
/Users/guillaumemulocher/.pyenv/versions/ansible-avd/lib/python3.10/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

vs

```
blah% python -W ignore::UserWarning -m pylint test.py

-------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 8.33/10, +1.67)
```

Trying to disable the warnings within the code is not working as pylint is not really executing it.

No solution so far except tweaking ansible-test somehow..

File: ansible_collections/arista/avd/roles/eos_designs/tasks/main.yml (hunk 3 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/tasks/main.yml]: well we don't want `debug, never` on the verify requirements. IMO we should just do this on every run unless it is very slow?

File: ansible_collections/arista/avd/plugins/action/verify_requirements.py (hunk 1 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/plugins/action/verify_requirements.py]: Removed pip for packaging

File: ansible_collections/arista/avd/plugins/action/verify_requirements.py [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+import os
+import sys
+from subprocess import PIPE, Popen
+
+import yaml
+from ansible.errors import AnsibleActionFail
+from ansible.module_utils.compat.importlib import import_module
+from ansible.plugins.action import ActionBase
+from ansible.utils.collection_loader._collection_finder import _get_collection_metadata
+from ansible.utils.display import Display
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/plugins/action/verify_requirements.py]: import the instantiated `display` like we do in schema-tools.

File: ansible_collections/arista/avd/plugins/README.md [hunk 5]
```diff
@@ -696,3 +696,142 @@ Example:
...
+            "micro": 3,
+            "releaselevel": "final",
+            "serial": 0
+        },
+        "python_path": [
+            "/Users/guillaumemulocher/.pyenv/versions/3.10.3/envs/ansible-avd/bin",
+            "/Users/guillaumemulocher/.pyenv/versions/3.10.3/lib/python310.zip",
+            "/Users/guillaumemulocher/.pyenv/versions/3.10.3/lib/python3.10",
+            "/Users/guillaumemulocher/.pyenv/versions/3.10.3/lib/python3.10/lib-dynload",
+            "/Users/guillaumemulocher/.pyenv/versions/3.10.3/envs/ansible-avd/lib/python3.10/site-packages"
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/README.md]:
```suggestion
            "/tmp/.pyenv/versions/3.10.3/envs/ansible-avd/bin",
            "/tmp/.pyenv/versions/3.10.3/lib/python310.zip",
            "/tmp/.pyenv/versions/3.10.3/lib/python3.10",
            "/tmp/.pyenv/versions/3.10.3/lib/python3.10/lib-dynload",
            "/tmp/.pyenv/versions/3.10.3/envs/ansible-avd/lib/python3.10/site-packages"
```

**Review by ClausHolbechArista — APPROVED** [review]

**Review by carlbuchmann — APPROVED** [review]
LGTM!
Tested in native environment, venv, pyenv and docker.

**Pull request merged**